core/src/main/scala/kafka/server/instrumentation/ProduceRequestInstrumentation.scala

```diff
@@ -0,0 +1,193 @@
+  val marks: mutable.Map[Stage.Value, Long] = mutable.Map()
+  var topicPartitionsInRequest: Iterable[TopicPartition] = List()
+
+  // Implicitly add init stage on construction
+  markStage(ProduceRequestInstrumentation.Stage.Init)
+
+  /**
+   * The core of the instrumentation.
+   *
+   * When the object is passed along, at any point in the code flow, it can invoke this method
+   * to mark the beginning of the stage with the current timestamp.
+   * The timestamps can later be used to calculate time taken in each stage via a diff on the 2 time stamps
+   *
+   * @param stage The stage to mark for beginning
+   */
+  def markStage(stage: Stage.Value): Unit = {
+    marks.put(stage, time.milliseconds())
+  }
+}
+object ProduceRequestInstrumentationLogger {
+  private val logger = Logger("produce.request.instrumentation.logger")
+  /**
+   * The config to control the instrumentation behavior
+   *
+   * @param thresholdToLogMs If the request completed within the threshold, don't log
+   * @param logRatio         Ratio of requests to log, applied after all funnels/thresholds
+   */
+  case class Config(thresholdToLogMs: Long,
+                    logRatio: Double)
+
+  val ReconfigurableConfigs: Set[String] = Set(
+    KafkaConfig.LiLongTailProduceRequestLogThresholdMsProp,
+    KafkaConfig.LiLongTailProduceRequestLogRatioProp,
+  )
+}
+
+/**
+ * Responsible for logging the instrumentation details.
+ * The config of the logger is associated with DynamicBrokerConfig lifecycle
+ *
+ * @param kafkaConfig For configuring dynamic broker config lifecycle
+ * @param replicaManager Used to obtain the # of partitions in PendingShrinkIsr
+ */
+class ProduceRequestInstrumentationLogger(kafkaConfig: KafkaConfig,
+                                          time: Time,
+                                          rnd: scala.util.Random,
+                                          replicaManager: ReplicaManager) extends Logging with BrokerReconfigurable {
+  override lazy val logger = ProduceRequestInstrumentationLogger.logger
+
+  // This is a cache of the dynamic config.
+  // Should be initialized/reconfigured within the BrokerReconfigurable lifecycle
+  validateReconfiguration(kafkaConfig)
+  kafkaConfig.dynamicConfig.addBrokerReconfigurable(this)
+  var config: Config = extractConfigFromKafkaConfig(kafkaConfig)
+
+  override def reconfigurableConfigs: collection.Set[String] = ProduceRequestInstrumentationLogger.ReconfigurableConfigs
+
+  override def validateReconfiguration(newConfig: KafkaConfig): Unit = {
+    val thresholdMs = newConfig.longTailProduceRequestLogThresholdMs
+    if (thresholdMs < -1) {
+      throw new ConfigException(s"${KafkaConfig.LiLongTailProduceRequestLogThresholdMsProp} should be >= -1")
+    }
+    val ratio = newConfig.longTailProduceRequestLogRatio
+    if (ratio < 0 || ratio > 1) {
+      throw new ConfigException(s"${KafkaConfig.LiLongTailProduceRequestLogRatioProp} should be between 0 to 1")
```

Comment:
I haven't checked upstream: is this going to kill the broker, or does it just get logged as an exception and then ignored? (I'm guessing the latter since it would be a massive DDoS screwup if a ZK mistake could wipe the entire cluster, but let's be sure.)

Reply:
Just checked --- the broker is quite resilient

On script side it rejects the update from
```
Error while executing config command with args '--bootstrap-server localhost:9092 --entity-type brokers --entity-default --alter --add-config li.instrumentation.requests.produce.long.tail.log.ratio=1000'
java.util.concurrent.ExecutionException: org.apache.kafka.common.errors.InvalidRequestException: Invalid config value for resource ConfigResource(type=BROKER, name=''): Invalid value 1000.0 for configuration li.instrumentation.requests.produce.long.tail.log.ratio: Value must be no more than 1
        at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:395)
        at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2022)
        at org.apache.kafka.common.internals.KafkaFutureImpl.get(KafkaFutureImpl.java:180)
        at kafka.admin.ConfigCommand$.alterConfig(ConfigCommand.scala:379)
        at kafka.admin.ConfigCommand$.processCommand(ConfigCommand.scala:327)
        at kafka.admin.ConfigCommand$.main(ConfigCommand.scala:98)
        at kafka.admin.ConfigCommand.main(ConfigCommand.scala)
Caused by: org.apache.kafka.common.errors.InvalidRequestException: Invalid config value for resource ConfigResource(type=BROKER, name=''): Invalid value 1000.0 for configuration li.instrumentation.requests.produce.long.tail.log.ratio: Value must be no more than 1
        at org.apache.kafka.common.protocol.Errors.exception(Errors.java:421)
        at org.apache.kafka.common.requests.ApiError.exception(ApiError.java:104)
        at org.apache.kafka.clients.admin.KafkaAdminClient$11.handleResponse(KafkaAdminClient.java:2455)
        at org.apache.kafka.clients.admin.KafkaAdminClient$AdminClientRunnable.handleResponses(KafkaAdminClient.java:1282)
        at org.apache.kafka.clients.admin.KafkaAdminClient$AdminClientRunnable.processRequests(KafkaAdminClient.java:1435)
        at org.apache.kafka.clients.admin.KafkaAdminClient$AdminClientRunnable.run(KafkaAdminClient.java:1358)
        at java.base/java.lang.Thread.run(Thread.java:829)
```

With a direct write to ZK:

1. The error will be thrown after the broker restarts, and loads the static configs
```
[2023-06-30 03:30:55,360] ERROR Dynamic default cluster config contains invalid values in: Set(li.instrumentation.requests.produce.long.tail.log.ratio), these configs will be ignored (kafka.server.DynamicBrokerConfig)
org.apache.kafka.common.config.ConfigException: Invalid value 1000.0 for configuration li.instrumentation.requests.produce.long.tail.log.ratio: Value must be no more than 1
```
2. On runtime force write to ZK wouldn't have any effect.

Reply:
OK, I have some concerns about the blocking-startup part (which could be a critical issue in mitigating an oncall issue), but given that we haven't managed to blow things up so far, maybe I'm worrying too much. I do see the benefits of causing a failure only on startup; silent failures can waste a lot of time.